File: NetFramework/FlexGrid/VB/AdvancedFiltering/InheritedFilterNewMenu.vb
Imports C1.Win.C1FlexGrid
Imports System
Imports System.Xml

Namespace AdvancedFiltering
    Friend Class InheritedFilterNewMenu
        Inherits ColumnFilter
        Public WeekdayFilter As WeekdayFilter = New WeekdayFilter()

        Public Overrides Function GetEditor() As IC1ColumnFilterEditor
            Return New InheritedFilterEditorNewMenu()
        End Function

        Public Overrides Sub Reset()
            MyBase.Reset()
            WeekdayFilter.Reset()
        End Sub

        Public Overrides Function Apply(ByVal value As Object) As Boolean
            Return MyBase.Apply(value) AndAlso WeekdayFilter.Apply(value)
        End Function

        Public Overrides ReadOnly Property IsActive As Boolean
            Get
                Return MyBase.IsActive OrElse WeekdayFilter.IsActive
            End Get
        End Property

        Public Overrides Sub WriteXmlDefinition(ByVal xw As XmlWriter)
            MyBase.WriteXmlDefinition(xw)

            If WeekdayFilter.IsActive Then
                xw.WriteAttributeString("ActiveFilter", WeekdayFilter.[GetType]().Name)
                xw.WriteStartElement(WeekdayFilter.[GetType]().Name)
                xw.WriteStartElement("Monday")
                xw.WriteAttributeString("Value", WeekdayFilter.Monday.ToString())
                xw.WriteEndElement()
                xw.WriteStartElement("Tuesday")
                xw.WriteAttributeString("Value", WeekdayFilter.Tuesday.ToString())
                xw.WriteEndElement()
                xw.WriteStartElement("Wednesday")
                xw.WriteAttributeString("Value", WeekdayFilter.Wednesday.ToString())
                xw.WriteEndElement()
                xw.WriteStartElement("Thursday")
                xw.WriteAttributeString("Value", WeekdayFilter.Thursday.ToString())
                xw.WriteEndElement()
                xw.WriteStartElement("Friday")
                xw.WriteAttributeString("Value", WeekdayFilter.Friday.ToString())
                xw.WriteEndElement()
                xw.WriteStartElement("Saturday")
                xw.WriteAttributeString("Value", WeekdayFilter.Saturday.ToString())
                xw.WriteEndElement()
                xw.WriteStartElement("Sunday")
                xw.WriteAttributeString("Value", WeekdayFilter.Sunday.ToString())
                xw.WriteEndElement()
                xw.WriteEndElement()
            End If
        End Sub

        Public Overrides Sub ReadXmlDefinition(ByVal xr As XmlReader)
            ReadXmlDefinition(xr, GetType(DateTime))

            While xr.Read()
                If xr.NodeType = XmlNodeType.Element Then
                    If Equals(xr.Name, "Monday") Then
                        Boolean.TryParse(xr.GetAttribute("Value"), WeekdayFilter.Monday)
                    ElseIf Equals(xr.Name, "Tuesday") Then
                        Boolean.TryParse(xr.GetAttribute("Value"), WeekdayFilter.Tuesday)
                    ElseIf Equals(xr.Name, "Wednesday") Then
                        Boolean.TryParse(xr.GetAttribute("Value"), WeekdayFilter.Wednesday)
                    ElseIf Equals(xr.Name, "Thursday") Then
                        Boolean.TryParse(xr.GetAttribute("Value"), WeekdayFilter.Thursday)
                    ElseIf Equals(xr.Name, "Friday") Then
                        Boolean.TryParse(xr.GetAttribute("Value"), WeekdayFilter.Friday)
                    ElseIf Equals(xr.Name, "Saturday") Then
                        Boolean.TryParse(xr.GetAttribute("Value"), WeekdayFilter.Saturday)
                    ElseIf Equals(xr.Name, "Sunday") Then
                        Boolean.TryParse(xr.GetAttribute("Value"), WeekdayFilter.Sunday)
                    End If
                End If
            End While
        End Sub
    End Class
End Namespace

File: NetFramework/FlexGrid/CS/AdvancedFiltering/InheritedFilterNewMenu.cs
using C1.Win.C1FlexGrid;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace AdvancedFiltering
{
    internal class InheritedFilterNewMenu : ColumnFilter
    {
        public WeekdayFilter WeekdayFilter = new WeekdayFilter();

        public override IC1ColumnFilterEditor GetEditor()
        {
            return new InheritedFilterEditorNewMenu();
        }

        public override void Reset()
        {
            base.Reset();
            WeekdayFilter.Reset();
        }

        public override bool Apply(object value)
        {
            return base.Apply(value) && WeekdayFilter.Apply(value);
        }

        public override bool IsActive
        {
            get
            {
                return base.IsActive || WeekdayFilter.IsActive;
            }
        }

        public override void WriteXmlDefinition(XmlWriter xw)
        {
            base.WriteXmlDefinition(xw);

            if (WeekdayFilter.IsActive)
            {
                xw.WriteAttributeString("ActiveFilter", WeekdayFilter.GetType().Name);
                xw.WriteStartElement(WeekdayFilter.GetType().Name);
                xw.WriteStartElement("Monday");
                xw.WriteAttributeString("Value", WeekdayFilter.Monday.ToString());
                xw.WriteEndElement();
                xw.WriteStartElement("Tuesday");
                xw.WriteAttributeString("Value", WeekdayFilter.Tuesday.ToString());
                xw.WriteEndElement();
                xw.WriteStartElement("Wednesday");
                xw.WriteAttributeString("Value", WeekdayFilter.Wednesday.ToString());
                xw.WriteEndElement();
                xw.WriteStartElement("Thursday");
                xw.WriteAttributeString("Value", WeekdayFilter.Thursday.ToString());
                xw.WriteEndElement();
                xw.WriteStartElement("Friday");
                xw.WriteAttributeString("Value", WeekdayFilter.Friday.ToString());
                xw.WriteEndElement();
                xw.WriteStartElement("Saturday");
                xw.WriteAttributeString("Value", WeekdayFilter.Saturday.ToString());
                xw.WriteEndElement();
                xw.WriteStartElement("Sunday");
                xw.WriteAttributeString("Value", WeekdayFilter.Sunday.ToString());
                xw.WriteEndElement();
                xw.WriteEndElement();
            }
        }

        public override void ReadXmlDefinition(XmlReader xr)
        {
            base.ReadXmlDefinition(xr);

            while (xr.Read())
            {
                if (xr.NodeType == XmlNodeType.Element)
                {
                    if (xr.Name == "Monday")
                    {
                        bool.TryParse(xr.GetAttribute("Value"), out WeekdayFilter.Monday);
                    }
                    else if (xr.Name == "Tuesday")
                    {
                        bool.TryParse(xr.GetAttribute("Value"), out WeekdayFilter.Tuesday);
                    }
                    else if (xr.Name == "Wednesday")
                    {
                        bool.TryParse(xr.GetAttribute("Value"), out WeekdayFilter.Wednesday);
                    }
                    else if (xr.Name == "Thursday")
                    {
                        bool.TryParse(xr.GetAttribute("Value"), out WeekdayFilter.Thursday);
                    }
                    else if (xr.Name == "Friday")
                    {
                        bool.TryParse(xr.GetAttribute("Value"), out WeekdayFilter.Friday);
                    }
                    else if (xr.Name == "Saturday")
                    {
                        bool.TryParse(xr.GetAttribute("Value"), out WeekdayFilter.Saturday);
                    }
                    else if (xr.Name == "Sunday")
                    {
                        bool.TryParse(xr.GetAttribute("Value"), out WeekdayFilter.Sunday);
                    }
                }
            }
        }
    }
}
